src/lib/comms/sol-coap.c

```diff
@@ -696,9 +740,14 @@ on_can_write(void *data, struct sol_socket *s)
     if (!outgoing)
         return false;
 
-    err = sol_socket_sendmsg(s, outgoing->pkt->buf.data,
-        outgoing->pkt->buf.used, &outgoing->cliaddr);
+    buf = prepare_buffer(outgoing);
```

Comment:
as per above, you can either return a buffer, instead of a pointer to a static, or receive a buffer in `prepare_buffer()`, then you give it your local struct to be filled.